// zcatalyst-sdk-node/lib/datastore/datastore.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var isValidApp = require('../utils/validator').isValidApp;
var isStringOrNumber = require('../utils/validator').isStringOrNumber;
var isNumber = require('../utils/validator').isNumber;
var wrapValidatorsWithPromise = require('../utils/validator').wrapValidatorsWithPromise;
var USER = require('../utils/constants').CREDENTIAL_USER;
var Table = require('./table').Table;
var CatalystAppInternals = require('../catalyst-app').CatalystAppInternals;
var AuthorizedHttpClient = require('../utils/api-request').AuthorizedHttpClient;
var API_VERSION = require('../utils/constants').API_VERSION;
var PRODUCT_NAME = require('../utils/constants').PRODUCT_NAME;
var CatalystDatastoreError = require('../utils/error').CatalystDatastoreError;
var Datastore = /** @class */ (function (_super) {
    __extends(Datastore, _super);
    function Datastore(app) {
        var _this = this;
        isValidApp(app, true);
        _this = _super.call(this, app.internal) || this;
        _this.requester = new AuthorizedHttpClient(_this);
        return _this;
    }
    Datastore.prototype.getAllTables = function () {
        var _this = this;
        _super.prototype.switchUser.call(this, USER.admin);
        var request = {
            method: 'GET',
            path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + this.projectId + "/table"
        };
        return this.requester.send(request).then(function (resp) {
            var json = resp.data;
            var tableArr = [];
            json.data.forEach(function (table) {
                tableArr.push(new Table(table, _this));
            });
            return tableArr;
        });
    };
    Datastore.prototype.getTableDetails = function (id) {
        var _this = this;
        _super.prototype.switchUser.call(this, USER.admin);
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'table_id', true);
        }, CatalystDatastoreError).then(function () {
            var request = {
                method: 'GET',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.projectId + "/table/" + id
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return new Table(json.data, _this);
            });
        });
    };
    Datastore.prototype.table = function (id) {
        if (!isStringOrNumber(id)) {
            throw new CatalystDatastoreError('invalid-argument', 'Value provided for table_id/table_name must be a non empty String or Number.', id);
        }
        if (isNumber(Number(id))) {
            return new Table({ table_id: id }, this);
        }
        else {
            return new Table({ table_name: id }, this);
        }
    };
    return Datastore;
}(CatalystAppInternals));
exports.Datastore = Datastore;
