// zcatalyst-sdk-node/lib/datastore/table.js
'use strict';
var isNonNullObject = require('../utils/validator').isNonNullObject;
var isStringOrNumber = require('../utils/validator').isStringOrNumber;
var isNonNullArray = require('../utils/validator').isNonNullArray;
var wrapValidatorsWithPromise = require('../utils/validator').wrapValidatorsWithPromise;
var USER = require('../utils/constants').CREDENTIAL_USER;
var API_VERSION = require('../utils/constants').API_VERSION;
var PRODUCT_NAME = require('../utils/constants').PRODUCT_NAME;
var CatalystDatastoreError = require('../utils/error').CatalystDatastoreError;
var Table = /** @class */ (function () {
    function Table(tableDetails, datastoreInstance) {
        this.superInstance = datastoreInstance;
        this._tableDetails = tableDetails;
        this.requester = datastoreInstance.requester;
    }
    Table.prototype.getColumnDetails = function (id) {
        var _this = this;
        this.superInstance.switchUser(USER.admin);
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'column_id', true);
        }, CatalystDatastoreError).then(function () {
            var request = {
                method: 'GET',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/table/" + (_this._tableDetails.table_id || _this._tableDetails.table_name) + "/column/" + id
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data;
            });
        });
    };
    Table.prototype.getAllColumns = function () {
        this.superInstance.switchUser(USER.admin);
        var request = {
            method: 'GET',
            path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + this.superInstance.projectId + "/table/" + (this._tableDetails.table_id || this._tableDetails.table_name) + "/column"
        };
        return this.requester.send(request).then(function (resp) {
            var json = resp.data;
            return json.data;
        });
    };
    Table.prototype.insertRow = function (row) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isNonNullObject(row, 'row', true);
        }, CatalystDatastoreError).then(function () {
            var rowArr = [];
            rowArr.push(row);
            var request = {
                method: 'POST',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/table/" + (_this._tableDetails.table_id || _this._tableDetails.table_name) + "/row",
                data: rowArr,
                type: 'json'
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data[0];
            });
        });
    };
    Table.prototype.insertRows = function (rowArr) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isNonNullArray(rowArr, 'rows', true);
        }, CatalystDatastoreError).then(function () {
            var request = {
                method: 'POST',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/table/" + (_this._tableDetails.table_id || _this._tableDetails.table_name) + "/row",
                data: rowArr,
                type: 'json'
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data;
            });
        });
    };
    Table.prototype.getAllRows = function () {
        this.superInstance.switchUser(USER.user);
        var request = {
            method: 'GET',
            path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + this.superInstance.projectId + "/table/" + (this._tableDetails.table_id || this._tableDetails.table_name) + "/row"
        };
        return this.requester.send(request).then(function (resp) {
            var json = resp.data;
            return json.data;
        });
    };
    Table.prototype.getRow = function (id) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'row_id', true);
        }, CatalystDatastoreError).then(function () {
            var request = {
                method: 'GET',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/table/" + (_this._tableDetails.table_id || _this._tableDetails.table_name) + "/row/" + id
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data;
            });
        });
    };
    Table.prototype.deleteRow = function (id) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'row_id', true);
        }, CatalystDatastoreError).then(function () {
            var request = {
                method: 'DELETE',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/table/" + (_this._tableDetails.table_id || _this._tableDetails.table_name) + "/row/" + id
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data ? true : false;
            });
        });
    };
    Table.prototype.updateRow = function (row) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isNonNullObject(row, 'row', true);
            isStringOrNumber(row.ROWID, 'ROWID in row', true);
        }, CatalystDatastoreError).then(function () {
            var request = {
                method: 'PUT',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/table/" + (_this._tableDetails.table_id || _this._tableDetails.table_name) + "/row",
                data: [row],
                type: 'json'
            };
            return _this.requester.send(request, _this).then(function (resp) {
                var json = resp.data;
                return json.data[0];
            });
        });
    };
    Table.prototype.updateRows = function (rows) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isNonNullArray(rows, 'row', true);
        }, CatalystDatastoreError).then(function () {
            var request = {
                method: 'PUT',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/table/" + (_this._tableDetails.table_id || _this._tableDetails.table_name) + "/row",
                data: rows,
                type: 'json'
            };
            return _this.requester.send(request, _this).then(function (resp) {
                var json = resp.data;
                return json.data;
            });
        });
    };
    Table.prototype.toString = function () {
        return JSON.stringify(this._tableDetails);
    };
    Table.prototype.toJSON = function () {
        return this._tableDetails;
    };
    return Table;
}());
exports.Table = Table;
