// zcatalyst-sdk-node/lib/zia/zia.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var isValidApp = require('../utils/validator').isValidApp;
var isNonEmptyString = require('../utils/validator').isNonEmptyString;
var wrapValidatorsWithPromise = require('../utils/validator').wrapValidatorsWithPromise;
var CatalystAppInternals = require('../catalyst-app').CatalystAppInternals;
var AuthorizedHttpClient = require('../utils/api-request').AuthorizedHttpClient;
var CatalystZiaError = require('../utils/error').CatalystZiaError;
var API_VERSION = require('../utils/constants').API_VERSION;
var PRODUCT_NAME = require('../utils/constants').PRODUCT_NAME;
var USER = require('../utils/constants').CREDENTIAL_USER;
var FORM = require('../utils/form-data');
var Zia = /** @class */ (function (_super) {
    __extends(Zia, _super);
    function Zia(app) {
        var _this = this;
        isValidApp(app, true);
        _this = _super.call(this, app.internal) || this;
        _super.prototype.switchUser.call(_this, USER.admin);
        _this.requester = new AuthorizedHttpClient(_this);
        return _this;
    }
    Zia.prototype.detectObject = function (file) {
        var formData = new FORM();
        formData.append('image', file);
        var request = {
            method: 'POST',
            path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + this.projectId + "/ml/detect-object",
            data: formData,
            type: 'file'
        };
        return this.requester.send(request).then(function (resp) {
            var json = resp.data;
            return json.data;
        });
    };
    Zia.prototype.extractOpticalCharacters = function (file, languageCodes) {
        var _this = this;
        return wrapValidatorsWithPromise(function () {
            isNonEmptyString(languageCodes, 'language_codes', true);
        }, CatalystZiaError).then(function () {
            var formData = new FORM();
            formData.append('image', file);
            var request = {
                method: 'POST',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.projectId + "/ml/ocr",
                data: formData,
                type: 'file'
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data.text;
            });
        });
    };
    Zia.prototype.extractNamedEntity = function (file) {
        var formData = new FORM();
        if (isNonEmptyString(file)) {
            formData.append('text', file);
        }
        else {
            formData.append('text_file', file);
        }
        var request = {
            method: 'POST',
            path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + this.projectId + "/ml/ner",
            data: formData,
            type: 'file'
        };
        return this.requester.send(request).then(function (resp) {
            var json = resp.data;
            return json.data;
        });
    };
    return Zia;
}(CatalystAppInternals));
exports.Zia = Zia;
