// zcatalyst-sdk-node/lib/push-notification/web-notification.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var wrapValidatorsWithPromise = require('../utils/validator').wrapValidatorsWithPromise;
var isNonEmptyString = require('../utils/validator').isNonEmptyString;
var isNonNullArray = require('../utils/validator').isNonNullArray;
var CatalystAppInternals = require('../catalyst-app').CatalystAppInternals;
var API_VERSION = require('../utils/constants').API_VERSION;
var PRODUCT_NAME = require('../utils/constants').PRODUCT_NAME;
var CatalystPushNotificationError = require('../utils/error').CatalystPushNotificationError;
var WebNotification = /** @class */ (function (_super) {
    __extends(WebNotification, _super);
    function WebNotification(notificationInstance) {
        var _this = _super.call(this, notificationInstance) || this;
        _this.requester = notificationInstance.requester;
        return _this;
    }
    WebNotification.prototype.sendNotification = function (message, recipients) {
        var _this = this;
        return wrapValidatorsWithPromise(function () {
            isNonEmptyString(message, 'message', true);
            isNonNullArray(recipients, 'recipients', true);
        }, CatalystPushNotificationError).then(function () {
            var request = {
                method: 'POST',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.projectId + "/project-user/notify",
                data: {
                    message: message,
                    recipients: recipients
                },
                type: 'json'
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data;
            });
        });
    };
    return WebNotification;
}(CatalystAppInternals));
exports.WebNotification = WebNotification;
