// zcatalyst-sdk-node/lib/search/search.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var isValidApp = require('../utils/validator').isValidApp;
var isNonNullObject = require('../utils/validator').isNonNullObject;
var validateObject = require('../utils/validator').validateObject;
var wrapValidatorsWithPromise = require('../utils/validator').wrapValidatorsWithPromise;
var CatalystAppInternals = require('../catalyst-app').CatalystAppInternals;
var AuthorizedHttpClient = require('../utils/api-request').AuthorizedHttpClient;
var API_VERSION = require('../utils/constants').API_VERSION;
var PRODUCT_NAME = require('../utils/constants').PRODUCT_NAME;
var CatalystSearchError = require('../utils/error').CatalystSearchError;
var Search = /** @class */ (function (_super) {
    __extends(Search, _super);
    function Search(app) {
        var _this = this;
        isValidApp(app, true);
        _this = _super.call(this, app.internal) || this;
        // super.switchUser(USER.admin);
        _this.requester = new AuthorizedHttpClient(_this);
        return _this;
    }
    Search.prototype.executeSearchQuery = function (searchQuery) {
        var _this = this;
        return wrapValidatorsWithPromise(function () {
            isNonNullObject(searchQuery, 'search_object', true);
            validateObject(searchQuery, ['search', 'search_table_columns'], 'search_object', true);
        }, CatalystSearchError).then(function () {
            var request = {
                method: 'POST',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.projectId + "/search",
                data: searchQuery,
                type: 'json'
            };
            return _this.requester.send(request).then(function (resp) {
                return resp.data.data;
            });
        });
    };
    return Search;
}(CatalystAppInternals));
exports.Search = Search;
