// zcatalyst-sdk-node/lib/file/folder.js
'use strict';
var isNonNullObject = require('../utils/validator').isNonNullObject;
var isStringOrNumber = require('../utils/validator').isStringOrNumber;
var isNonEmptyString = require('../utils/validator').isNonEmptyString;
var wrapValidatorsWithPromise = require('../utils/validator').wrapValidatorsWithPromise;
var USER = require('../utils/constants').CREDENTIAL_USER;
var API_VERSION = require('../utils/constants').API_VERSION;
var PRODUCT_NAME = require('../utils/constants').PRODUCT_NAME;
var CatalystFilestoreError = require('../utils/error').CatalystFilestoreError;
var FORM = require('../utils/form-data');
var Folder = /** @class */ (function () {
    function Folder(folderDetails, fileInstance) {
        this.superInstance = fileInstance;
        this._folderDetails = folderDetails;
        this.requester = fileInstance.requester;
    }
    Folder.prototype.update = function (folderDetails) {
        var _this = this;
        this.superInstance.switchUser(USER.admin);
        return wrapValidatorsWithPromise(function () {
            isNonNullObject(folderDetails, 'folder_object', true);
            isNonEmptyString(folderDetails.folder_name, 'folder_name in folder_object', true);
        }, CatalystFilestoreError).then(function () {
            var postData = {
                folder_name: folderDetails.folder_name
            };
            var request = {
                method: 'PUT',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/folder/" + _this._folderDetails.id,
                data: postData,
                type: 'json'
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data;
            });
        });
    };
    Folder.prototype.delete = function () {
        this.superInstance.switchUser(USER.admin);
        var request = {
            method: 'DELETE',
            path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + this.superInstance.projectId + "/folder/" + this._folderDetails.id
        };
        return this.requester.send(request).then(function (resp) {
            var json = resp.data;
            if (json.data) {
                return true;
            }
        });
    };
    Folder.prototype.getFileDetails = function (id) {
        var _this = this;
        this.superInstance.switchUser(USER.admin);
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'file_id', true);
        }, CatalystFilestoreError).then(function () {
            var request = {
                method: 'GET',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/folder/" + _this._folderDetails.id + "/file/" + id
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data;
            });
        });
    };
    Folder.prototype.deleteFile = function (id) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'file_id', true);
        }, CatalystFilestoreError).then(function () {
            var request = {
                method: 'DELETE',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/folder/" + _this._folderDetails.id + "/file/" + id
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                if (json.data) {
                    return true;
                }
            });
        });
    };
    Folder.prototype.uploadFile = function (fileDetails) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isNonNullObject(fileDetails, 'file_object', true);
            isNonEmptyString(fileDetails.name, 'name in file_object', true);
        }, CatalystFilestoreError).then(function () {
            var formData = new FORM();
            formData.append('code', fileDetails.code);
            formData.append('file_name', fileDetails.name);
            var request = {
                method: 'POST',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/folder/" + _this._folderDetails.id + "/file",
                data: formData,
                type: 'file'
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return json.data;
            });
        });
    };
    Folder.prototype.downloadFile = function (id) {
        var _this = this;
        this.superInstance.switchUser(USER.user);
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'file_id', true);
        }, CatalystFilestoreError).then(function () {
            var request = {
                method: 'GET',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.superInstance.projectId + "/folder/" + _this._folderDetails.id + "/file/" + id + "/download"
            };
            return _this.requester.send(request).then(function (resp) {
                return resp.buffer;
            });
        });
    };
    Folder.prototype.toString = function () {
        return JSON.stringify(this._folderDetails);
    };
    Folder.prototype.toJSON = function () {
        return this._folderDetails;
    };
    return Folder;
}());
exports.Folder = Folder;
