// zcatalyst-sdk-node/lib/queue/queue.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var isValidApp = require('../utils/validator').isValidApp;
var isStringOrNumber = require('../utils/validator').isStringOrNumber;
var wrapValidatorsWithPromise = require('../utils/validator').wrapValidatorsWithPromise;
var Topic = require('./topic').Topic;
var CatalystAppInternals = require('../catalyst-app').CatalystAppInternals;
var AuthorizedHttpClient = require('../utils/api-request').AuthorizedHttpClient;
var API_VERSION = require('../utils/constants').API_VERSION;
var PRODUCT_NAME = require('../utils/constants').PRODUCT_NAME;
var USER = require('../utils/constants').CREDENTIAL_USER;
var CatalystQueueError = require('../utils/error').CatalystQueueError;
var Queue = /** @class */ (function (_super) {
    __extends(Queue, _super);
    function Queue(app) {
        var _this = this;
        isValidApp(app, true);
        _this = _super.call(this, app.internal) || this;
        _super.prototype.switchUser.call(_this, USER.admin);
        _this.requester = new AuthorizedHttpClient(_this);
        return _this;
    }
    Queue.prototype.getAllTopics = function () {
        var _this = this;
        var request = {
            method: 'GET',
            path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + this.projectId + "/queue"
        };
        return this.requester.send(request).then(function (resp) {
            var json = resp.data;
            var topicsArr = [];
            json.data.forEach(function (topic) {
                topicsArr.push(new Topic(topic, _this));
            });
            return topicsArr;
        });
    };
    Queue.prototype.getTopicDetails = function (id) {
        var _this = this;
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'queue_id', true);
        }, CatalystQueueError).then(function () {
            var request = {
                method: 'GET',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.projectId + "/queue/" + id
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return new Topic(json.data, _this);
            });
        });
    };
    Queue.prototype.topic = function (id) {
        if (!isStringOrNumber(id)) {
            throw new CatalystQueueError('invalid-argument', 'Value provided for topic_id must be a non empty String or Number.', id);
        }
        return new Topic({ id: id }, this);
    };
    return Queue;
}(CatalystAppInternals));
exports.Queue = Queue;
