// zcatalyst-sdk-node/lib/cache/cache.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var PRODUCT_NAME = require('../utils/constants').PRODUCT_NAME;
var API_VERSION = require('../utils/constants').API_VERSION;
var USER = require('../utils/constants').CREDENTIAL_USER;
var Segment = require('./segment').Segment;
var isValidApp = require('../utils/validator').isValidApp;
var isStringOrNumber = require('../utils/validator').isStringOrNumber;
var wrapValidatorsWithPromise = require('../utils/validator').wrapValidatorsWithPromise;
var CatalystAppInternals = require('../catalyst-app').CatalystAppInternals;
var AuthorizedHttpClient = require('../utils/api-request').AuthorizedHttpClient;
var CatalystCacheError = require('../utils/error').CatalystCacheError;
var Cache = /** @class */ (function (_super) {
    __extends(Cache, _super);
    function Cache(app) {
        var _this = this;
        isValidApp(app, true);
        _this = _super.call(this, app.internal) || this;
        _super.prototype.switchUser.call(_this, USER.admin);
        _this.requester = new AuthorizedHttpClient(_this);
        return _this;
    }
    Cache.prototype.getAllSegment = function () {
        var _this = this;
        var request = {
            method: 'GET',
            path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + this.projectId + "/segment"
        };
        return this.requester.send(request).then(function (resp) {
            var json = resp.data;
            var segmentsArr = [];
            json.data.forEach(function (segment) {
                segmentsArr.push(new Segment(segment, _this));
            });
            return segmentsArr;
        });
    };
    Cache.prototype.getSegmentDetails = function (id) {
        var _this = this;
        return wrapValidatorsWithPromise(function () {
            isStringOrNumber(id, 'segment_id', true);
        }, CatalystCacheError).then(function () {
            var request = {
                method: 'GET',
                path: "/" + PRODUCT_NAME + "/" + API_VERSION + "/project/" + _this.projectId + "/segment/" + id
            };
            return _this.requester.send(request).then(function (resp) {
                var json = resp.data;
                return new Segment(json.data, _this);
            });
        });
    };
    Cache.prototype.segment = function (id) {
        var segmentId = typeof id !== 'undefined' ? id : 'default';
        if (!isStringOrNumber(segmentId)) {
            throw new CatalystCacheError('invalid-argument', 'Provided value segment_id must be a string or number', segmentId);
        }
        return new Segment({ id: segmentId }, this);
    };
    return Cache;
}(CatalystAppInternals));
exports.Cache = Cache;
