// zcatalyst-sdk-node/lib/utils/error.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var CatalystError = /** @class */ (function (_super) {
    __extends(CatalystError, _super);
    function CatalystError(errorInfo) {
        var _this = _super.call(this, errorInfo.message) || this;
        _this.errorInfo = errorInfo;
        return _this;
    }
    Object.defineProperty(CatalystError.prototype, "code", {
        /** @return {string} The error code. */
        get: function () {
            return this.errorInfo.code;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CatalystError.prototype, "message", {
        /** @return {string} The error message. */
        get: function () {
            return this.errorInfo.message;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CatalystError.prototype, "value", {
        /** @return {any} The value that caused this error. */
        get: function () {
            return this.errorInfo.value;
        },
        enumerable: false,
        configurable: true
    });
    /** @return {object} The object representation of the error. */
    CatalystError.prototype.toJSON = function () {
        return {
            code: this.code,
            message: this.message
        };
    };
    /** @return {string} The string representation of the error. */
    CatalystError.prototype.toString = function () {
        return JSON.stringify(this.toJSON());
    };
    return CatalystError;
}(Error));
exports.CatalystError = CatalystError;
var PrefixedCatalystError = /** @class */ (function (_super) {
    __extends(PrefixedCatalystError, _super);
    function PrefixedCatalystError(codePrefix, code, message, value) {
        var _this = _super.call(this, {
            code: codePrefix + "/" + code,
            message: message,
            value: value
        }) || this;
        _this.codePrefix = codePrefix;
        return _this;
    }
    return PrefixedCatalystError;
}(CatalystError));
exports.PrefixedCatalystError = PrefixedCatalystError;
var CatalystApiError = /** @class */ (function (_super) {
    __extends(CatalystApiError, _super);
    function CatalystApiError(code, message, value) {
        return _super.call(this, 'api', code, message, value) || this;
    }
    return CatalystApiError;
}(PrefixedCatalystError));
exports.CatalystApiError = CatalystApiError;
var CatalystAppError = /** @class */ (function (_super) {
    __extends(CatalystAppError, _super);
    function CatalystAppError(code, message, value) {
        return _super.call(this, 'app', code, message, value) || this;
    }
    return CatalystAppError;
}(PrefixedCatalystError));
exports.CatalystAppError = CatalystAppError;
var CatalystAuthError = /** @class */ (function (_super) {
    __extends(CatalystAuthError, _super);
    function CatalystAuthError(code, message, value) {
        return _super.call(this, 'auth', code, message, value) || this;
    }
    return CatalystAuthError;
}(PrefixedCatalystError));
exports.CatalystAuthError = CatalystAuthError;
var CatalystCacheError = /** @class */ (function (_super) {
    __extends(CatalystCacheError, _super);
    function CatalystCacheError(code, message, value) {
        return _super.call(this, 'cache', code, message, value) || this;
    }
    return CatalystCacheError;
}(PrefixedCatalystError));
exports.CatalystCacheError = CatalystCacheError;
var CatalystCronError = /** @class */ (function (_super) {
    __extends(CatalystCronError, _super);
    function CatalystCronError(code, message, value) {
        return _super.call(this, 'cron', code, message, value) || this;
    }
    return CatalystCronError;
}(PrefixedCatalystError));
exports.CatalystCronError = CatalystCronError;
var CatalystDatastoreError = /** @class */ (function (_super) {
    __extends(CatalystDatastoreError, _super);
    function CatalystDatastoreError(code, message, value) {
        return _super.call(this, 'datastore', code, message, value) || this;
    }
    return CatalystDatastoreError;
}(PrefixedCatalystError));
exports.CatalystDatastoreError = CatalystDatastoreError;
var CatalystEmailError = /** @class */ (function (_super) {
    __extends(CatalystEmailError, _super);
    function CatalystEmailError(code, message, value) {
        return _super.call(this, 'email', code, message, value) || this;
    }
    return CatalystEmailError;
}(PrefixedCatalystError));
exports.CatalystEmailError = CatalystEmailError;
var CatalystFilestoreError = /** @class */ (function (_super) {
    __extends(CatalystFilestoreError, _super);
    function CatalystFilestoreError(code, message, value) {
        return _super.call(this, 'filestore', code, message, value) || this;
    }
    return CatalystFilestoreError;
}(PrefixedCatalystError));
exports.CatalystFilestoreError = CatalystFilestoreError;
var CatalystFunctionsError = /** @class */ (function (_super) {
    __extends(CatalystFunctionsError, _super);
    function CatalystFunctionsError(code, message, value) {
        return _super.call(this, 'functions', code, message, value) || this;
    }
    return CatalystFunctionsError;
}(PrefixedCatalystError));
exports.CatalystFunctionsError = CatalystFunctionsError;
var CatalystGQLError = /** @class */ (function (_super) {
    __extends(CatalystGQLError, _super);
    function CatalystGQLError(code, message, value) {
        return _super.call(this, 'gql', code, message, value) || this;
    }
    return CatalystGQLError;
}(PrefixedCatalystError));
exports.CatalystGQLError = CatalystGQLError;
var CatalystPushNotificationError = /** @class */ (function (_super) {
    __extends(CatalystPushNotificationError, _super);
    function CatalystPushNotificationError(code, message, value) {
        return _super.call(this, 'push-notification', code, message, value) || this;
    }
    return CatalystPushNotificationError;
}(PrefixedCatalystError));
exports.CatalystPushNotificationError = CatalystPushNotificationError;
var CatalystQueueError = /** @class */ (function (_super) {
    __extends(CatalystQueueError, _super);
    function CatalystQueueError(code, message, value) {
        return _super.call(this, 'queue', code, message, value) || this;
    }
    return CatalystQueueError;
}(PrefixedCatalystError));
exports.CatalystQueueError = CatalystQueueError;
var CatalystSearchError = /** @class */ (function (_super) {
    __extends(CatalystSearchError, _super);
    function CatalystSearchError(code, message, value) {
        return _super.call(this, 'search', code, message, value) || this;
    }
    return CatalystSearchError;
}(PrefixedCatalystError));
exports.CatalystSearchError = CatalystSearchError;
var CatalystUserManagementError = /** @class */ (function (_super) {
    __extends(CatalystUserManagementError, _super);
    function CatalystUserManagementError(code, message, value) {
        return _super.call(this, 'user-management', code, message, value) || this;
    }
    return CatalystUserManagementError;
}(PrefixedCatalystError));
exports.CatalystUserManagementError = CatalystUserManagementError;
var CatalystZCQLError = /** @class */ (function (_super) {
    __extends(CatalystZCQLError, _super);
    function CatalystZCQLError(code, message, value) {
        return _super.call(this, 'zcql', code, message, value) || this;
    }
    return CatalystZCQLError;
}(PrefixedCatalystError));
exports.CatalystZCQLError = CatalystZCQLError;
var CatalystZiaError = /** @class */ (function (_super) {
    __extends(CatalystZiaError, _super);
    function CatalystZiaError(code, message, value) {
        return _super.call(this, 'zia', code, message, value) || this;
    }
    return CatalystZiaError;
}(PrefixedCatalystError));
exports.CatalystZiaError = CatalystZiaError;
var CatalystConnectorError = /** @class */ (function (_super) {
    __extends(CatalystConnectorError, _super);
    function CatalystConnectorError(code, message, value) {
        return _super.call(this, 'connector', code, message, value) || this;
    }
    return CatalystConnectorError;
}(PrefixedCatalystError));
exports.CatalystConnectorError = CatalystConnectorError;
